### PR TITLE
Add GetDeckPath to OutputReader

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
In order to cache the deck file, we need to add a new attribute `DeckPath` to `InMemoryOutputReader`

Related https://github.com/flyteorg/flytepropeller/pull/443

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2175

## Follow-up issue
_NA_

### CODE REVIEW

**Review by EngHabu — APPROVED** [review]

**Comment by EngHabu** [comment]
Looks like you may need to add a few more unit tests for patch coverage...

File: go/tasks/plugins/array/outputs.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func (w assembleOutputsWorker) Process(ctx context.Context, workItem workqueue.W
 	}
 
 	ow := ioutils.NewRemoteFileOutputWriter(ctx, i.dataStore, i.outputPaths)
-	if err = ow.Put(ctx, ioutils.NewInMemoryOutputReader(finalOutputs, nil)); err != nil {
+	if err = ow.Put(ctx, ioutils.NewInMemoryOutputReader(finalOutputs, nil, nil)); err != nil {
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply go/tasks/plugins/array/outputs.go:110]: okay, will create a new issue for it.

**Review by EngHabu — APPROVED** [review]

**Comment by honnix** [comment]
This is unfortunately a breaking change that we could have been more verbose and bumped minor (even major) version. One of our plugins needs to be updated to incorporate this. What is deck? Can we pass a `nil` if our plugin doesn't care?

**Comment by honnix** [comment]
> What is deck? Can we pass a `nil` if our plugin doesn't care?

OK I found it here https://docs.flyte.org/projects/cookbook/en/latest/auto/core/flyte_basics/deck.html#sphx-glr-auto-core-flyte-basics-deck-py. Seem to be quite useful. So how we do pass in the path instead of `nil`?

**Comment by pingsutw** [comment]
@honnix what error did you run into? you can set deckPath to nil in `NewInMemoryOutputReader`.
In flytekit, `output.pb`  will be written to `output_dir/output.pb`and Flyte deck (an HTML file) will be written to `output_dir/deck.html`

https://github.com/pingsutw/flytepropeller/blob/f08ccaad75246c7a1149be632206c0278608e61e/pkg/controller/nodes/executor.go#L239-L249

Some plugins (like bigquery) that will not generate a deck file, we also set deckPath to nil.

**Comment by honnix** [comment]
Thanks for replying @pingsutw .

> what error did you run into?

Since the signature of `NewInMemoryOutputReader` was changed by adding one more argument, one of our internal plugins failed to compile. Since this type of APIs could be used by any plugin, we might want to be more careful when making changes in the future.

The same as you suggested, we put nil as the deck path to fix it, and now everything is fine.

**Comment by pingsutw** [comment]
@honnix got it, sorry. we'll be more careful when making changes to public API in the future.

**Comment by honnix** [comment]
@pingsutw No worries. It was easy to catch and fix. :)

**Pull request merged**